File: server/internal/domain/strategy/contract.go
package strategy

import (
	"encoding/json"

	"github.com/shopspring/decimal"

	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/common"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/indicator"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/market"
	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/sentiment"
)

type Params map[string]any

type Definition struct {
	StrategyID        string                    `json:"strategy_id"`
	Version           string                    `json:"version"`
	Family            string                    `json:"family,omitempty"`
	ParametersSchema  json.RawMessage           `json:"parameters_schema"`
	InputRequirements []string                  `json:"input_requirements"`
	OverlayTypes      []string                  `json:"overlay_types"`
	WarmUpCandles     func(Params) (int, error) `json:"-"`
	IsComposite       bool                      `json:"is_composite"`
	DisplayName       string                    `json:"display_name"`
	Description       string                    `json:"description"`
	CodeFingerprint   string                    `json:"code_fingerprint,omitempty"`
}

type AnalysisContext struct {
	Provider      string
	Symbol        string
	Timeframe     common.Timeframe
	Candles       market.CausalCandles
	Index         int
	Indicators    indicator.IndicatorView
	NewsSentiment *sentiment.NewsSentimentWindow
	Params        Params
}

type Signal struct {
	Action     common.Action    `json:"action"`
	Confidence *decimal.Decimal `json:"confidence,omitempty"`
	Price      *decimal.Decimal `json:"price,omitempty"`
	SignedSize *decimal.Decimal `json:"signed_size,omitempty"`
	Evidence   json.RawMessage  `json:"evidence,omitempty"`
}

type Strategy interface {
	Definition() Definition
	Analyze(AnalysisContext) (Signal, error)
}

type Reference struct {
	StrategyID string `json:"strategy_id"`
	Version    string `json:"version"`
}

type ChildSignal struct {
	Strategy Reference       `json:"strategy"`
	Signal   Signal          `json:"signal"`
	Weight   decimal.Decimal `json:"weight"`
}

type ResolvedSignal struct {
	StrategyID string
	Version    string
	Signal     Signal
	Weight     decimal.Decimal
}

type CombinationPolicy struct {
	Policy    string          `json:"policy"`
	Threshold decimal.Decimal `json:"threshold"`
	Encoding  string          `json:"encoding"`
}

type ChildDefinition struct {
	StrategyID string          `json:"strategy_id"`
	Version    string          `json:"version"`
	Parameters Params          `json:"parameters"`
	Weight     decimal.Decimal `json:"weight"`
}

type CompositeDefinition struct {
	StrategyID  string            `json:"strategy_id"`
	Version     string            `json:"version"`
	Children    []ChildDefinition `json:"children"`
	Combination CombinationPolicy `json:"combination"`
}

type CompositeSpec = CompositeDefinition

type CandidateStrategy struct {
	Definition     CompositeDefinition `json:"definition"`
	CandidateHash  string               `json:"candidate_hash"`
	GeneratedBy    string               `json:"generated_by"`
	GenerationMeta json.RawMessage      `json:"generation_meta,omitempty"`
}

File: app/domain/strategy/contract.py
"""Strategy domain contracts (float64).

Mirrors `server/internal/domain/strategy/contract.go`. All numeric fields are
Python `float` (float64).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol

from ..common import Action, Decimal, Timeframe
from ..indicator import IndicatorView
from ..market import CausalCandles
from ..sentiment import NewsSentimentWindow

Params = dict[str, Any]


@dataclass
class Definition:
    strategy_id: str
    version: str
    family: str | None = None
    parameters_schema: Any | None = None
    input_requirements: list[str] = field(default_factory=list)
    overlay_types: list[str] = field(default_factory=list)
    warm_up_candles: Any | None = None  # callable(Params) -> int
    is_composite: bool = False
    display_name: str = ""
    description: str = ""
    code_fingerprint: str | None = None


@dataclass
class AnalysisContext:
    provider: str
    symbol: str
    timeframe: Timeframe
    candles: CausalCandles
    index: int
    indicators: IndicatorView
    news_sentiment: NewsSentimentWindow | None = None
    params: Params = field(default_factory=dict)


@dataclass
class Signal:
    action: Action
    confidence: Decimal | None = None
    price: Decimal | None = None
    signed_size: Decimal | None = None
    evidence: Any | None = None


class Strategy(Protocol):
    def definition(self) -> Definition: ...

    def analyze(self, context: AnalysisContext) -> Signal: ...


@dataclass
class Reference:
    strategy_id: str
    version: str


@dataclass
class ChildSignal:
    strategy: Reference
    signal: Signal
    weight: Decimal


@dataclass
class ResolvedSignal:
    strategy_id: str
    version: str
    signal: Signal
    weight: Decimal


@dataclass
class CombinationPolicy:
    policy: str
    threshold: Decimal
    encoding: dict[str, int]


@dataclass
class ChildDefinition:
    strategy_id: str
    version: str
    parameters: Params
    weight: Decimal


@dataclass
class CompositeDefinition:
    strategy_id: str
    version: str
    children: list[ChildDefinition]
    combination: CombinationPolicy


CompositeSpec = CompositeDefinition


@dataclass
class CandidateStrategy:
    definition: CompositeDefinition
    candidate_hash: str
    generated_by: str
    generation_meta: Any | None = None
